### PR TITLE
🌱 Enable more gocritic lint checks

### LINKED ISSUE #4419: Add linter to prevent use of deprecated ioutil module
**User Story**

As a developer, I would like to make sure I am following guidelines and avoid using deprecated code.

Go 1.16 deprecated the `ioutil` module, adding `os` and `io` equivalents for included functions.

https://github.com/kubernetes-sigs/cluster-api/pull/4399#issuecomment-811504262 updated the cluster-api code to use those alternative calls.

The `ioutil` deprecation can be considered a soft-deprecation as it does not trigger deprecation errors from linters, presumably to make the transition easier for consumers of the deprecated code. Eventually this is likely to become a hard deprecation and be removed.

Since we've cleaned up existing usage, it would be good to make sure new instances are not introduced as people contribute new code. This will make it easier when that hard deprecation comes.

**Detailed Description**

See comments in https://github.com/kubernetes-sigs/cluster-api/pull/4399#issuecomment-811504262 for some suggested approaches to add linting checks to prevent the introduction of new uses of the module.

/kind feature

### PR DESCRIPTION
**Pull request by stmcginnis**
**What this PR does / why we need it**:

There are a multiple groups of checks run by gocritic. Some are enabled
by default, but others require explicit enabling. This enables the
experimental group, minus a few exceptions that would either be too
disruptive or are things we aren't concerned about.

One of the main benefits of this extra coverage is the flagging of using
the deprecated ioutil package.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #4419

### CODE REVIEW

File: controllers/topology/internal/extensions/patches/inline/json_patch_generator.go [hunk 1]
```diff
@@ -347,7 +347,7 @@ func parsePathSegment(segment string) (*pathSegment, error) {
 	}
 
 	return &pathSegment{
-		path:  segment[:strings.Index(segment, leftArrayDelim)], //nolint:gocritic // We already check above that segment contains leftArrayDelim,
+		path:  segment[:strings.Index(segment, leftArrayDelim)], //nolint:gocritic
```
Comment (sbueringer) [inline controllers/topology/internal/extensions/patches/inline/json_patch_generator.go:350]: Can we keep the comment?

File: test/infrastructure/docker/main.go [hunk 2]
```diff
@@ -104,6 +104,9 @@ func initFlags(fs *pflag.FlagSet) {
 
 func main() {
 	rand.Seed(time.Now().UnixNano())
+	if _, err := os.ReadDir("/tmp/"); err != nil {
+		panic(err)
+	}
```
Comment (sbueringer) [inline test/infrastructure/docker/main.go:109]:
```suggestion
```
:)

**Review by sbueringer — COMMENTED** [review]
Thx!
2 nits otherwise lgtm

File: test/infrastructure/docker/internal/provisioning/ignition/kindadapter.go [hunk 3]
```diff
@@ -88,7 +88,7 @@ func getActions(userData []byte) ([]provisioning.Cmd, error) {
 
 	for _, u := range ignition.Systemd.Units {
 		contents := strings.TrimSpace(u.Contents)
-		path := filepath.Join("/etc/systemd/system", u.Name)
+		path := fmt.Sprintf("/etc/systemd/system/%s", u.Name)
```
Comment (invidian) [inline test/infrastructure/docker/internal/provisioning/ignition/kindadapter.go:91]: Why is this change necessary? Shouldn't `filepath.Join` be used when dealing with joining paths?

Comment (sbueringer) [reply test/infrastructure/docker/internal/provisioning/ignition/kindadapter.go:91]: I think filepath.Join doesn't really help as the separator is already hard-coded in the first path (but not sure if that's the idea behind the finding)

Comment (stmcginnis) [reply test/infrastructure/docker/internal/provisioning/ignition/kindadapter.go:91]: Yep, along with what @sbueringer says, it doesn't make sense in this context because we never need to account for Windows path separators. There will never be a case where we are running on an OS where the path would need to be `\etc\systemd\system` rather than `/etc/systemd/system`.

File: test/infrastructure/docker/main.go (hunk 2 above)
Comment (stmcginnis) [reply test/infrastructure/docker/main.go:109]: Sorry, what is the suggested change here?

File: controllers/topology/internal/extensions/patches/inline/json_patch_generator.go (hunk 1 above)
Comment (stmcginnis) [reply controllers/topology/internal/extensions/patches/inline/json_patch_generator.go:350]: Honestly can't remember why this was dropped in the first place. :D

File: test/infrastructure/docker/internal/provisioning/ignition/kindadapter.go (hunk 3 above)
Comment (invidian) [reply test/infrastructure/docker/internal/provisioning/ignition/kindadapter.go:91]: Fair :+1:

File: test/infrastructure/docker/main.go (hunk 2 above)
Comment (stmcginnis) [reply test/infrastructure/docker/main.go:109]: I'm guessing you meant to change the panic to `os.exit(1)` to match the other exits from main. Update coming with that, but please let me know if there is a different point I've missed. Thanks!

**Review by killianmuldoon — APPROVED** [review]
/lgtm

**Review by vincepri — APPROVED** [review]
/approve
